cms_coding_rules.md

```diff
@@ -0,0 +1,144 @@
+#### [7 -- Design and Coding Guidelines](#7----design-and-coding-guidelines-1)
+
+## 1 -- Introduction
+This document describes the CMS C++ software naming, coding, style, and documentation rules and recommendations.
+All CMS C++ software is expected to comply with the rules. The asterisk (\*) after some rules indicates that there may be exceptional use cases where the rule may be violated with good justiﬁcation.
+Coding rules are meant to prevent serious problems in software function, performance, maintainability, usability, and portability. The Packaging Rules section also has some brief guidelines for Python configuration.
+
+## 2 -- Naming Rules
+1. C++ header ﬁles use the sufﬁx .h, e.g. `CaloCluster.h`. (\*)
+2. For C++ source ﬁles, the preferred sufﬁx  is .cc, e.g. `CaloCluster.cc`, though there are existing files using .cpp and .cxx suffixes. (\*)
+3. For a header file that contains a class, name that ﬁle after the class.
+4. Name source ﬁles after the class.
+5. For class, struct, type, and enumeration names use upper class initials, e.g. `GeometryBuilder`.
+6. For namespaces use lower case, e.g. `namespace edm`.
+7. Start method names with lowercase, use upper case initials for following words, e.g. `collisionPoint()`.  
+  Allowed exception: Implementation of virtual methods inherited from external packages, e.g. `ProcessHits()` method required by Geant4.  
+8. Start data member names with lower case. A trailing "\_" is the preferred method to distinguish a data member from the getter method (e.g. `momentum_`).
+9. Using "set" for a setter method is preferred, e.g. `setMomentum(double momentum)`.
+10. For a getter method, using the value name is preferred, e.g. `momentum()`.
+11. Do not use single character names, except for loop indices.
+12. Do not use special characters, except for "_" where allowed.
+13. Do not use "\_" as ﬁrst character. Only use it as the last character for class data member names, not local variable names.
+14. Do not use "\_\_".
+15. Use clear and explanatory variable names.
+
+## 3 -- Style Rules
+1. Do not indent pre-processor directives -- there should be no leading spaces before a directive.  (\*)
+2. Never change the language syntax using `#deﬁne`.
+3. Do not use spaces between a function, constructor, or method name and its opening parenthesis, e.g. `energy()` rather than `energy ()`. A similar style is encouraged but not required when brace initialization is used, e.g. `energy{13}`.
+4. Do not use spaces in front of [] or on either side of -> . For example, `position[i]` instead of `position [i]`.
+5. Separate expressions in a `for` statement by spaces.
+6. Use the same indentation for comments as for the block the comments refer to.
+
+## 4 -- Technical Coding Rules
+1. Protect each header ﬁle from multiple inclusion with:  
+`#ifndef PackageName_SubPackageName_FileName_h`  
+`#define PackageName_SubPackageName_FileName_h`  
+(body of header file)  
+`#endif`  
```

Comment:
I like to use and suggest
`#endif // PackageName_SubPackageName_FileName_h`

Reply:
Added.

Reply:
Could we leave this as optional? I don't really care about this enough to recommend anyone spend time removing these extra comments or lose any sleep if someone likes them and adds them. But to me the meaning of a #endif at the end of a header file is obvious and this comment does not help one understand the code when reading it. It is just extra boilerplate to ignore that distracts from the important code, also it is just one more thing to remember to add in when creating a header file, then one has to make sure it is consistent with the top of the file, and there is a new way to make an error. Comments at the end of blocks of code are helpful when it is not obvious what is being ended, but otherwise it just seems like clutter to me.

Reply:
On second thought, I think I agree with @wddgit. Anything in code that can diverge will diverge, and we will end up with trailing endif comments that don't match their ifndefs and will only confuse people.
I will wait until tomorrow to see if there are more comments about this point.

Reply:
In the interests of simplicity, I removed the additional comment from the \#endif. Such comments are not prohibited, but they are not required either.